### PR TITLE
Support generating typescript cells and pages | Autodetect ts project

### LINKED ISSUE #523: TypeScript: Generators and Templates tracking issue
_This is a sub-tracking issue of #234_ 

>This is a component of Redwood's initiative to add first-class TypeScript support. 🎉

Generators are a part of the CLI package and located in this directory:
https://github.com/redwoodjs/redwood/tree/master/packages/cli/src/commands/generate

The CLI package is built with Yargs. Read more [here](https://github.com/redwoodjs/redwood/blob/master/packages/cli/src/commands/generate/README.md).

## Goal
The intention is to provide a path for a _collaborative, incremental_ transition to having 100% TypeScript generators. This means the scope of a PR can be an individual generator, which can be merged into the master branch on its own when it's ready.

- **source templates will be `.ts`**
- **generators should be aware of both `.ts` and `.js`**; this applies to 1) refusing to run if files already exist and 2) overwriting existing files using `--force`
    - In the case where existing files are a _different_ language target (e.g. in the case of a hybrid-project), the generator should refuse to run regardless of using the `--force` option; a clear explanation message should be given
- **templates should be transpiled and formatted when the target language is JS** (suggested to use `ts.transpile` and prettier, but this needs evaluation)
    - need to determine appropriate `ts.ScriptTarget`; "ESNext" seems like a candidate
    - ⚠️transpilation formatting goal == _match existing JS templates_ so that no updates are required for Redwood Tutorial, either code snippets or images (TBD as needed)
- **generators should have CLI options for targeting `--javascript` or `--typescript`**
    - generators should not _require_ a language target option; instead, the default target will be the project "language default" or a config override in `redwood.toml` if given.
    -  Example: if `tsconfig.json` exists and there is no `redwood.toml` override, the generator will automatically set `--typescript` to `true`. Therefore, running `yarn rw g [option]` will target typescript. 
    - TODO: create a new Internal package helper for easy access to this config value
- **generators need to pass CI tests _and_ the generated `...test.ts|js` files need to pass `yarn rw test` and `yarn rw lint`**

## Steps to Convert a Specific Generator
1. Rewrite each generator template, including template test file, as TS
    - TS files will _replace_ JS files
2. Convert corresponding `__tests__` directory files and fixtures to TS
3. Add a JS and TS target CLI command option, e.g. `[--javascript|--typescript]` to each generator CLI command
    - commands should also be aliased `[-js|-ts]` respectively
    - set the behavior to automatically use the project "language default" (see above) OR a given `--javascript` or `--typescript` option
4. When language target=javascript, transpile and format TS templates to JS
- do generated `..test.ts/js` files pass `yarn rw lint` and `yarn rw test`?

## Checklist: Templates and Generators
- [x] **Destroy** (New command in progress via #487)
    - [ ] confirm support for both ts and js filetypes, as well as hybrid Apps
    - [ ] add support as/if needed
- [ ] **Helpers** (~~PR #557~~ stalled out)
    - [ ] convert `__tests__/helpers.test.js` to TS
    - [ ] convert `helpers.js` to TS
- [ ] **src/lib Files**
    _note: these files are outside generators/_, but the functionality that relates to generators should be moved to `src/commands/generators`
    - [ ] src/lib/colors.js
    - [ ] src/lib/index.js
    - [ ] src/lib/test.js
- [x] **Component** #632
    - [x] convert `templates/` files to TS
    - [x] convert `__tests__/` fixtures and `test.js` to TS
    - [x] generator support for JS transpilation; enable `--javascript` as default
- [x] **Function**
    - [x] convert `templates/` file to TS
    - [x] convert `__tests__/` fixtures and `test.js` to TS
    - [x] generator support for JS transpilation; enable `--javascript` as default 
- [x] **Layout** #685 
    - [ ] convert `templates/` files to TS
    - [ ] convert `__tests__/` fixtures and `test.js` to TS
    - [ ] generator support for JS transpilation; enable `--javascript` as default 
- [ ] **Page**
⚠️ requires updates to Router to support TS
    - [ ] convert `templates/` files to TS
    - [ ] convert `__tests__/` fixtures and `test.js` to TS
    - [ ] generator support for JS transpilation; enable `--javascript` as default 
- [x] **SDL** 👉PR #515 
    - [x] convert `templates/` file to TS
    - [x] convert `__tests__/` fixtures and `test.js` to TS
    - [x] generator support for JS transpilation; enable `--javascript` as default 
- [x] **Service** 👉PR #515 
    - [x] convert `templates/` files to TS
    - [x] convert `__tests__/` fixtures and `test.js` to TS
    - [x] generator support for JS transpilation; enable `--javascript` as default 
- [ ] **Cell**
⚠️ requires additional support to replicate functionality in TS. Details in [this comment](https://github.com/redwoodjs/redwood/issues/234#issuecomment-620906084).
    - [ ] convert `templates/` files to TS
    - [ ] convert `__tests__/` fixtures and `test.js` to TS
    - [ ] generator support for JS transpilation; enable `--javascript` as default 
- [ ] **Scaffold** 
⚠️ requires SDL, Service, and Cells TS conversion
⚠️ requires updates to Router to support TS
    - [ ] convert `templates/` files to TS: components, layouts, and pages
    - [ ] convert `__tests__/` fixtures and `test.js` to TS
    - [ ] generator support for JS transpilation; enable `--javascript` as default

### PR DESCRIPTION
**Pull request by dac09**
Closes #523

## What does it do?
- Allows generation of cells in typescript
```yarn rw g cell pets --ts```
- Allows generation of pages in typescript
```yarn rw g page kitten --ts```

- Allows generation of scaffolds in typescript

- Autodetects if your project is typescript, and flags all generators to use typescript. So no need to pass the `--ts` flag all the time. In your typescript project you can just run `yarn rw g page kitten` and it'll infer the ts flag

- Updates all the tests to use `toMatchSnapshot`, and cleans up their behaviour so all the handlers are consistent.

- Removes depracated generators (been long enough since Sep 2020!) 

### What it doesn't do
Statically type cells, or handle typing cells from the outside. This is a separate piece that Peter P is looking at.

### CODE REVIEW

File: packages/cli/src/commands/generate/layout/__tests__/layout.test.ts [hunk 1]
```diff
@@ -158,12 +165,22 @@ test("doesn't include test file when --tests is set to false", () => {
   ])
 })
 
-test.only('includes skip link when --skipLink is set to true', () => {
```
Comment (dac09) [inline packages/cli/src/commands/generate/layout/__tests__/layout.test.ts:161]: Uhoh @jtoar 😆

File: packages/cli/src/lib/index.js [hunk 2]
```diff
@@ -459,8 +459,12 @@ export const runCommandTask = async (commands, { verbose }) => {
  * Extract default CLI args from an exported builder
  */
 export const getDefaultArgs = (builder) => {
-  return Object.entries(builder).reduce((agg, [k, v]) => {
-    agg[k] = v.default
-    return agg
-  }, {})
+  return Object.entries(builder).reduce(
```
Comment (dac09) [inline packages/cli/src/lib/index.js:462]: Just renames the variables to make it more readable

File: packages/cli/src/commands/generate.js [hunk 3]
```diff
@@ -18,23 +23,19 @@ export const builder = (yargs) =>
...
-    description: 'Generate JavaScript files',
-    type: 'boolean',
-  },
   typescript: {
     alias: 'ts',
-    default: false,
-    description: 'Generate TypeScript files',
+    default: project.isTypeScriptProject,
+    description:
+      'Generate TypeScript files. Enabled by default if we detect your project is typescript',
```
Comment (dthyresson) [inline packages/cli/src/commands/generate.js:38]: > if we detect your project is typescript

Is that done in toml settings? Or inferred in some other fashion?

Comment (dac09) [reply packages/cli/src/commands/generate.js:38]: Uses @redwoodjs/structure. Just checks for tsconfig.json files really!

**Review by peterp — APPROVED** [review]
This is a monster... Thanks for doing this! 🦾, it looks good to me, but I didn't spend as much time as I would've wanted one this, but given the size and the fact that it's all tested makes me feel more confident.

**Comment by dac09** [comment]
I'll make sure to test these manually before the release with DT's help, but I was fairly careful - made sure all the tests pass first, then moved them to snapshots, then checked each individual snapshot

**Pull request merged**